clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -377,18 +377,26 @@ synchronized void incrementSequenceNumber(TopicPartition topicPartition, int inc
         sequenceNumbers.put(topicPartition, currentSequenceNumber);
     }
 
-    synchronized TxnRequestHandler nextRequestHandler() {
+    synchronized TxnRequestHandler nextRequestHandler(boolean hasIncompleteBatches) {
         if (!newPartitionsInTransaction.isEmpty())
             enqueueRequest(addPartitionsToTransactionHandler());
 
-        TxnRequestHandler nextRequestHandler = pendingRequests.poll();
-        if (nextRequestHandler != null && maybeTerminateRequestWithError(nextRequestHandler)) {
+        TxnRequestHandler nextRequestHandler = pendingRequests.peek();
+        if (nextRequestHandler == null)
+            return null;
+
+        // Do not send the EndTxn until all batches have been flushed
+        if (nextRequestHandler.isEndTxn() && hasIncompleteBatches)
```

Comment:
In addition to what jason said, we also fail any future sends with an IllegalStateException if they are called when we are in any state but the `IN_TRANSACTION` state. This protects against a badly written multithreaded application which tries to do sends from another thread while being blocked on the `commit` call on one thread.

So there can be on change in the queued batches after a `commitTransaction` or `abortTransaction` is called.